# iFood CRM Data Analyst Case
---
Rafael Faria Castelão

## Importing and cleaning data

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime

In [2]:
# Import data
pilot_data = pd.read_csv('ml_project1_data.csv', index_col=0)

# Print information for dataset
print("Dataset contains: \n• {} training examples (rows)\n• {} features (columns)."
      .format(pilot_data.shape[0], pilot_data.shape[1]))

Dataset contains: 
• 2240 training examples (rows)
• 28 features (columns).


In [27]:
# Since data has 28 features let's set an option to display all features
pd.set_option('display.max_columns', 50)

Let's explore the data a little bit:

In [4]:
# Print a sample of the data
pilot_data.head()

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
ID                                                                        
5524        1957  Graduation         Single  58138.0        0         0   
2174        1954  Graduation         Single  46344.0        1         1   
4141        1965  Graduation       Together  71613.0        0         0   
6182        1984  Graduation       Together  26646.0        1         0   
5324        1981         PhD        Married  58293.0        1         0   

     Dt_Customer  Recency  MntWines  MntFruits  MntMeatProducts  \
ID                                                                
5524  2012-09-04       58       635         88              546   
2174  2014-03-08       38        11          1                6   
4141  2013-08-21       26       426         49              127   
6182  2014-02-10       26        11          4               20   
5324  2014-01-19       94       173         43              118   

      MntFishProducts  MntSweetProducts  MntGoldProds  NumDealsPurchases  \
ID                                                                         
5524              172                88            88                  3   
2174                2                 1             6                  2   
4141              111                21            42                  1   
6182               10                 3             5                  2   
5324               46                27            15                  5   

      NumWebPurchases  NumCatalogPurchases  NumStorePurchases  \
ID                                                              
5524                8                   10                  4   
2174                1                    1                  2   
4141                8                    2                 10   
6182                2                    0                  4   
5324                5                    3                  6   

      NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
ID                                                                  
5524                  7             0             0             0   
2174                  5             0             0             0   
4141                  4             0             0             0   
6182                  6             0             0             0   
5324                  5             0             0             0   

      AcceptedCmp1  AcceptedCmp2  Complain  Z_CostContact  Z_Revenue  Response  
ID                                                                              
5524             0             0         0              3         11         1  
2174             0             0         0              3         11         0  
4141             0             0         0              3         11         0  
6182             0             0         0              3         11         0  
5324             0             0         0              3         11         0

It's possible to see that the dataset has a mix of quantitative and categorial data.

In [5]:
# Print a summary of the data
pilot_data.describe()

Year_Birth         Income      Kidhome     Teenhome      Recency  \
count  2240.000000    2216.000000  2240.000000  2240.000000  2240.000000   
mean   1968.805804   52247.251354     0.444196     0.506250    49.109375   
std      11.984069   25173.076661     0.538398     0.544538    28.962453   
min    1893.000000    1730.000000     0.000000     0.000000     0.000000   
25%    1959.000000   35303.000000     0.000000     0.000000    24.000000   
50%    1970.000000   51381.500000     0.000000     0.000000    49.000000   
75%    1977.000000   68522.000000     1.000000     1.000000    74.000000   
max    1996.000000  666666.000000     2.000000     2.000000    99.000000   

          MntWines    MntFruits  MntMeatProducts  MntFishProducts  \
count  2240.000000  2240.000000      2240.000000      2240.000000   
mean    303.935714    26.302232       166.950000        37.525446   
std     336.597393    39.773434       225.715373        54.628979   
min       0.000000     0.000000         0.000000         0.000000   
25%      23.750000     1.000000        16.000000         3.000000   
50%     173.500000     8.000000        67.000000        12.000000   
75%     504.250000    33.000000       232.000000        50.000000   
max    1493.000000   199.000000      1725.000000       259.000000   

       MntSweetProducts  MntGoldProds  NumDealsPurchases  NumWebPurchases  \
count       2240.000000   2240.000000        2240.000000      2240.000000   
mean          27.062946     44.021875           2.325000         4.084821   
std           41.280498     52.167439           1.932238         2.778714   
min            0.000000      0.000000           0.000000         0.000000   
25%            1.000000      9.000000           1.000000         2.000000   
50%            8.000000     24.000000           2.000000         4.000000   
75%           33.000000     56.000000           3.000000         6.000000   
max          263.000000    362.000000          15.000000        27.000000   

       NumCatalogPurchases  NumStorePurchases  NumWebVisitsMonth  \
count          2240.000000        2240.000000        2240.000000   
mean              2.662054           5.790179           5.316518   
std               2.923101           3.250958           2.426645   
min               0.000000           0.000000           0.000000   
25%               0.000000           3.000000           3.000000   
50%               2.000000           5.000000           6.000000   
75%               4.000000           8.000000           7.000000   
max              28.000000          13.000000          20.000000   

       AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
count   2240.000000   2240.000000   2240.000000   2240.000000   2240.000000   
mean       0.072768      0.074554      0.072768      0.064286      0.013393   
std        0.259813      0.262728      0.259813      0.245316      0.114976   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

          Complain  Z_CostContact  Z_Revenue     Response  
count  2240.000000         2240.0     2240.0  2240.000000  
mean      0.009375            3.0       11.0     0.149107  
std       0.096391            0.0        0.0     0.356274  
min       0.000000            3.0       11.0     0.000000  
25%       0.000000            3.0       11.0     0.000000  
50%       0.000000            3.0       11.0     0.000000  
75%       0.000000            3.0       11.0     0.000000  
max       1.000000            3.0       11.0     1.000000

The first thing that I see here is that there is only one column with some missing values: **Income**. It has 24 values missing, so we will have to deal with this later.

Another funny aspect is that there are two features that I do not see in the meta-data table in the case file: **Z_CostContact** and **Z_Revenue**. Checking the means and standard deviation for these two features, it's possible to see that they have the same value for all the training examples. We can conclude these features will not be useful to the model, so we can drop them.

In [6]:
# Drop the columns Z_CostContact and Z_Revenue
pilot_data.drop(columns=['Z_CostContact', 'Z_Revenue'], inplace=True)

Like said before there are some categorical features in the dataset that need to be treated for the model to have a good fit.

In [7]:
# Checking for categorical data
print("Categorical features:\n{}".format(list(pilot_data.dtypes[(pilot_data.dtypes == 'object')].index)))

Categorical features:
['Education', 'Marital_Status', 'Dt_Customer']


Two of the features listed above are actually nominal features: **Education** and **Marital_Status**.

However, **Dt_Customer** the <ins>date</ins> of the customer's enrollment with the company. It is possible to redefine this feature to a continuous variable that represents the number of days this customer's enrolled with the company.

In [8]:
# Redefine Dt_Customer to represent the number of days the customer's enrolled with the company up to now
pilot_data['Dt_Customer'] = pilot_data['Dt_Customer'].map(lambda x: (datetime.now() - datetime.strptime(x, '%Y-%m-%d')).days)

In [9]:
pilot_data.head()

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
ID                                                                        
5524        1957  Graduation         Single  58138.0        0         0   
2174        1954  Graduation         Single  46344.0        1         1   
4141        1965  Graduation       Together  71613.0        0         0   
6182        1984  Graduation       Together  26646.0        1         0   
5324        1981         PhD        Married  58293.0        1         0   

      Dt_Customer  Recency  MntWines  MntFruits  MntMeatProducts  \
ID                                                                 
5524         2872       58       635         88              546   
2174         2322       38        11          1                6   
4141         2521       26       426         49              127   
6182         2348       26        11          4               20   
5324         2370       94       173         43              118   

      MntFishProducts  MntSweetProducts  MntGoldProds  NumDealsPurchases  \
ID                                                                         
5524              172                88            88                  3   
2174                2                 1             6                  2   
4141              111                21            42                  1   
6182               10                 3             5                  2   
5324               46                27            15                  5   

      NumWebPurchases  NumCatalogPurchases  NumStorePurchases  \
ID                                                              
5524                8                   10                  4   
2174                1                    1                  2   
4141                8                    2                 10   
6182                2                    0                  4   
5324                5                    3                  6   

      NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
ID                                                                  
5524                  7             0             0             0   
2174                  5             0             0             0   
4141                  4             0             0             0   
6182                  6             0             0             0   
5324                  5             0             0             0   

      AcceptedCmp1  AcceptedCmp2  Complain  Response  
ID                                                    
5524             0             0         0         1  
2174             0             0         0         0  
4141             0             0         0         0  
6182             0             0         0         0  
5324             0             0         0         0

Back to the other two categorical variables, **Education** and **Marital_Status** they have the following values:

In [10]:
print("Education: {} \nMarital Status: {}"
      .format(list(pilot_data['Education'].unique()), list(pilot_data['Marital_Status'].unique())))

Education: ['Graduation', 'PhD', 'Master', 'Basic', '2n Cycle'] 
Marital Status: ['Single', 'Together', 'Married', 'Divorced', 'Widow', 'Alone', 'Absurd', 'YOLO']


Both of these features are <ins>nominal variables</ins>, which means they do not have an intrinsic ranking. Since both of them have a small number of values, the approach we'll use for the model to deal with these features is the <ins>One-Hot Encoding</ins>.

P.S. I loved 'YOLO' marital status.

In [11]:
from sklearn.preprocessing import OneHotEncoder

In [12]:
# Apply One-Hot Encoding to each nominal feature
OH_encoder = OneHotEncoder(handle_unknown='ignore', sparse=False)
OH_cols = pd.DataFrame(OH_encoder.fit_transform(pilot_data[['Education', 'Marital_Status']]))

# Put the index in Hot-One Encoding
OH_cols.index = pilot_data.index

# Create a dictionary to rename the columns in a more friendly way, instead of numbers from 0 to 12
OH_columns_rename_dict = {k: v for (k,v) in 
                          zip(list(OH_cols.columns), 
                              list(OH_encoder.get_feature_names(['Education', 'Marital_Status'])))}

# Rename the columns for the One-Hot Encoding
OH_cols.rename(columns=OH_columns_rename_dict, inplace=True)

Just to have a glance on what Hot-Encoding does, these are the rewritten **Education** and **Marital_Status** features.

In [13]:
OH_cols.head()

Education_2n Cycle  Education_Basic  Education_Graduation  \
ID                                                                
5524                 0.0              0.0                   1.0   
2174                 0.0              0.0                   1.0   
4141                 0.0              0.0                   1.0   
6182                 0.0              0.0                   1.0   
5324                 0.0              0.0                   0.0   

      Education_Master  Education_PhD  Marital_Status_Absurd  \
ID                                                             
5524               0.0            0.0                    0.0   
2174               0.0            0.0                    0.0   
4141               0.0            0.0                    0.0   
6182               0.0            0.0                    0.0   
5324               0.0            1.0                    0.0   

      Marital_Status_Alone  Marital_Status_Divorced  Marital_Status_Married  \
ID                                                                            
5524                   0.0                      0.0                     0.0   
2174                   0.0                      0.0                     0.0   
4141                   0.0                      0.0                     0.0   
6182                   0.0                      0.0                     0.0   
5324                   0.0                      0.0                     1.0   

      Marital_Status_Single  Marital_Status_Together  Marital_Status_Widow  \
ID                                                                           
5524                    1.0                      0.0                   0.0   
2174                    1.0                      0.0                   0.0   
4141                    0.0                      1.0                   0.0   
6182                    0.0                      1.0                   0.0   
5324                    0.0                      0.0                   0.0   

      Marital_Status_YOLO  
ID                         
5524                  0.0  
2174                  0.0  
4141                  0.0  
6182                  0.0  
5324                  0.0

In [16]:
# Remove the nominal variables Education and Marital_Status
pilot_data.drop(['Education', 'Marital_Status'], axis=1, inplace=True, errors='ignore')

# Add One-Hot encoded columns to numerical features
pilot_data = pd.concat([pilot_data, OH_cols], axis=1)

This is how the dataset is up to now.

In [17]:
pilot_data.head()

Year_Birth   Income  Kidhome  Teenhome  Dt_Customer  Recency  MntWines  \
ID                                                                             
5524        1957  58138.0        0         0         2872       58       635   
2174        1954  46344.0        1         1         2322       38        11   
4141        1965  71613.0        0         0         2521       26       426   
6182        1984  26646.0        1         0         2348       26        11   
5324        1981  58293.0        1         0         2370       94       173   

      MntFruits  MntMeatProducts  MntFishProducts  MntSweetProducts  \
ID                                                                    
5524         88              546              172                88   
2174          1                6                2                 1   
4141         49              127              111                21   
6182          4               20               10                 3   
5324         43              118               46                27   

      MntGoldProds  NumDealsPurchases  NumWebPurchases  ...  Response  \
ID                                                      ...             
5524            88                  3                8  ...         1   
2174             6                  2                1  ...         0   
4141            42                  1                8  ...         0   
6182             5                  2                2  ...         0   
5324            15                  5                5  ...         0   

      Education_2n Cycle  Education_Basic  Education_Graduation  \
ID                                                                
5524                 0.0              0.0                   1.0   
2174                 0.0              0.0                   1.0   
4141                 0.0              0.0                   1.0   
6182                 0.0              0.0                   1.0   
5324                 0.0              0.0                   0.0   

      Education_Master  Education_PhD  Marital_Status_Absurd  \
ID                                                             
5524               0.0            0.0                    0.0   
2174               0.0            0.0                    0.0   
4141               0.0            0.0                    0.0   
6182               0.0            0.0                    0.0   
5324               0.0            1.0                    0.0   

      Marital_Status_Alone  Marital_Status_Divorced  Marital_Status_Married  \
ID                                                                            
5524                   0.0                      0.0                     0.0   
2174                   0.0                      0.0                     0.0   
4141                   0.0                      0.0                     0.0   
6182                   0.0                      0.0                     0.0   
5324                   0.0                      0.0                     1.0   

      Marital_Status_Single  Marital_Status_Together  Marital_Status_Widow  \
ID                                                                           
5524                    1.0                      0.0                   0.0   
2174                    1.0                      0.0                   0.0   
4141                    0.0                      1.0                   0.0   
6182                    0.0                      1.0                   0.0   
5324                    0.0                      0.0                   0.0   

      Marital_Status_YOLO  
ID                         
5524                  0.0  
2174                  0.0  
4141                  0.0  
6182                  0.0  
5324                  0.0  

[5 rows x 37 columns]

The last step to have a good dataset to fit the model is to deal with the missing values. Like we've seen before, the only column that presents missing values is the **Income** one.

Since this feature seems an important one, usually the household's income is used as a base for several predictions, it will not be a smart move to drop this whole column. Besides that, there are only 24 out of 2240 missing values.

A better approach is to use the imputation method, which fills in the missing values with some number. In this case, we will fill in the mean value along this column. 

In [18]:
from sklearn.impute import SimpleImputer

In [42]:
# Apply Simple Imputer to each column
my_imputer = SimpleImputer()
imputed_pilot_data = pd.DataFrame(my_imputer.fit_transform(pilot_data))

# Put the index and columns in Hot-One Encoding
imputed_pilot_data.index = pilot_data.index
imputed_pilot_data.columns = pilot_data.columns

# Let's use the pilot_data variable again
pilot_data = imputed_pilot_data.copy()

After all these steps, we finally have a good dataset to work with, in which the model will be able to fit well.

The final dataset looks like the following.

In [45]:
pilot_data.head()

Year_Birth   Income  Kidhome  Teenhome  Dt_Customer  Recency  MntWines  \
ID                                                                             
5524      1957.0  58138.0      0.0       0.0       2872.0     58.0     635.0   
2174      1954.0  46344.0      1.0       1.0       2322.0     38.0      11.0   
4141      1965.0  71613.0      0.0       0.0       2521.0     26.0     426.0   
6182      1984.0  26646.0      1.0       0.0       2348.0     26.0      11.0   
5324      1981.0  58293.0      1.0       0.0       2370.0     94.0     173.0   

      MntFruits  MntMeatProducts  MntFishProducts  MntSweetProducts  \
ID                                                                    
5524       88.0            546.0            172.0              88.0   
2174        1.0              6.0              2.0               1.0   
4141       49.0            127.0            111.0              21.0   
6182        4.0             20.0             10.0               3.0   
5324       43.0            118.0             46.0              27.0   

      MntGoldProds  NumDealsPurchases  NumWebPurchases  NumCatalogPurchases  \
ID                                                                            
5524          88.0                3.0              8.0                 10.0   
2174           6.0                2.0              1.0                  1.0   
4141          42.0                1.0              8.0                  2.0   
6182           5.0                2.0              2.0                  0.0   
5324          15.0                5.0              5.0                  3.0   

      NumStorePurchases  NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  \
ID                                                                       
5524                4.0                7.0           0.0           0.0   
2174                2.0                5.0           0.0           0.0   
4141               10.0                4.0           0.0           0.0   
6182                4.0                6.0           0.0           0.0   
5324                6.0                5.0           0.0           0.0   

      AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  Complain  Response  \
ID                                                                   
5524           0.0           0.0           0.0       0.0       1.0   
2174           0.0           0.0           0.0       0.0       0.0   
4141           0.0           0.0           0.0       0.0       0.0   
6182           0.0           0.0           0.0       0.0       0.0   
5324           0.0           0.0           0.0       0.0       0.0   

      Education_2n Cycle  Education_Basic  Education_Graduation  \
ID                                                                
5524                 0.0              0.0                   1.0   
2174                 0.0              0.0                   1.0   
4141                 0.0              0.0                   1.0   
6182                 0.0              0.0                   1.0   
5324                 0.0              0.0                   0.0   

      Education_Master  Education_PhD  Marital_Status_Absurd  \
ID                                                             
5524               0.0            0.0                    0.0   
2174               0.0            0.0                    0.0   
4141               0.0            0.0                    0.0   
6182               0.0            0.0                    0.0   
5324               0.0            1.0                    0.0   

      Marital_Status_Alone  Marital_Status_Divorced  Marital_Status_Married  \
ID                                                                            
5524                   0.0                      0.0                     0.0   
2174                   0.0                      0.0                     0.0   
4141                   0.0                      0.0                     0.0   
6182                   0.0                      0.0                  

## Fitting the Model

The first step to build a model is breaking the dataset into input and output variables. As the meta-data table in the case file states, **Response** variable is the target, that is what we want to predict. Since I believe all the other variables are important, or even slightly useful features, I'll define all other features as the input variables.

In [64]:
# Create the target vector
y = pilot_data['Response']

# Create a list of the input variables
pilot_features = list(set(pilot_data.columns) - set(['Response']))

# Create the input matrix
X = pilot_data[pilot_features]

The next step is to split data into training and validation sets.

The <ins>training data</ins> is the one used to fit the model, while the <ins>validation data</ins> is used to calculate the model quality.

In [70]:
from sklearn.model_selection import train_test_split

In [80]:
# Split data into training and validation data for both features and target
# Training data with 80% the size of the dataset
train_X, val_X, train_y, val_y = train_test_split(X, y, random_state = 0, train_size=0.8)

The next step is to understand which model we are using to fit the data.

Since this a supervised learning we are dealing with two types of algorithms: regression or classification. The ideia is to create a hypothesis function that will predict the value of y. Since y is a discrete output, we are using a classification algorithm.

In [86]:
print("Response (target) values:\n{}".format(list(y.unique())))

Response (target) values:
[1.0, 0.0]


P.S. As we can see above, the output value is a discrete variable with values of 1 or 0. It configures a binary classification problem.